diff -urN uClibc-0.9.30.1/extra/scripts/unifdef.c uClibc-0.9.30.1.fixed/extra/scripts/unifdef.c
--- uClibc-0.9.30.1/extra/scripts/unifdef.c	2008-04-27 18:10:00.000000000 -0500
+++ uClibc-0.9.30.1.fixed/extra/scripts/unifdef.c	2009-07-26 01:04:05.125883554 -0500
@@ -206,7 +206,7 @@
 static void             error(const char *);
 static int              findsym(const char *);
 static void             flushline(bool);
-static Linetype         getline(void);
+static Linetype         parseline(void);
 static Linetype         ifeval(const char **);
 static void             ignoreoff(void);
 static void             ignoreon(void);
@@ -512,7 +512,7 @@
 
 	for (;;) {
 		linenum++;
-		lineval = getline();
+		lineval = parseline();
 		trans_table[ifstate[depth]][lineval]();
 		debug("process %s -> %s depth %d",
 		    linetype_name[lineval],
@@ -526,7 +526,7 @@
  * help from skipcomment().
  */
 static Linetype
-getline(void)
+parseline(void)
 {
 	const char *cp;
 	int cursym;
